Dockerfile

```diff
@@ -1,9 +1,18 @@
-FROM drujensen/crystal:0.23.1
+FROM crystallang/crystal:0.23.1
```

Comment:
I'm torn here.  Should we maintain a crystal image since the official one takes forever to get updated?

Reply:
The official crystal image takes forever to get updated and it's _massive_. I went down the path of trying to get crystal built under a much more minimal image when I built [my amber docker image](https://github.com/robacarp/amber_build/blob/master/Dockerfile) and ran too far into the weeds to continue, but I'd love to see an amber image that isn't so ginormous.

Reply:
Actually, along those lines, a 2 stage docker file which builds amber and only copies the runtime dependencies into the final image, omitting the source code, would prevent the image from growing as the repo history grows.

Reply:
@robacarp yeah, I noticed the crystal image is pretty big.  I think its including everything to build crystal including llvm. Also crystallang/crystal:0.24.1 hasn't been released yet.  This held us up last time as well.  I will change this back to my image.

Regarding the 2 stage docker file, For development, It's something @elorest and I discussed and sounds like we should consider it.

For production, I think `crystal` and `amber` don't need to exist on the image.  An alpine flavor linux base image and only the application binary produced with the dependent files would be needed.  Maybe we should consider doing this for the user with the `deploy` command?

Reply:
@drujensen In building my production image, I found that I needed the amber binary to run migrations. Is that actually an issue? Or is there a better way to run migrations through the compiled app binary?

Reply:
Tangentially, a crystal image based off of alpine is what I tried to build and ran into packages needed to build the compiler not being quickly available. LLVM, I think.

Reply:
> In building my production image, I found that I needed the amber binary to run migration

Interesting point.   For production at work, I have been pushing to have the main web application service to be minimalistic and immutable.   

https://diogomonica.com/2016/11/19/increasing-attacker-cost-using-immutable-infrastructure/

For migrations, I create a temporary service that performs the migration and then exits.  I guess that could use the `amber` image to run the migration? Another option is we provide the DDL from the migrations that can be ran manually. Not sure.  I haven't spent much time thinking about a production solution for Amber but I think it's good to start that conversation.

Regarding Alpine, yes, very minimalistic.  Perfect for production images if you can figure it out ;-).  I don't doubt you would struggle to get crystal on it. After we compile the application, we no longer need crystal and amber, or anyway, I'm hopeful that is true.

Reply:
Hm, I'm pushing to a dokku host (self hosted heroku-ish) which runs the same container for migrations. That might not be the only way to do it with dokku, and I'm not sure what heroku  needs to run migrations. @faustinoaq do you know?

The crecto migration files are essentially SQL and, while that does cause some problems with unspoken expectations about what an orm is expecting data-type wise, even a simple bash script is capable of chopping the file and running the SQL. Knowing which part of which migration to run is a little more subtle.

At the very least, dockerhub makes it really easy to cascade-rebuild an image when one of its dependencies changes so making multiple variant child images which have or don't have the  build stack, migration tools, etc wouldn't be hard.

Reply:
> I'm not sure what heroku needs to run migrations. @faustinoaq do you know?

@eliasjpr implemented migration on [heroku buildpack for amber](https://github.com/amberframework/heroku-buildpack-amber/pull/3)

Reply:
@robacarp to run migrations on heroku it just needs to be specified in the Procfile in the `release` stage see https://amberframework.org/guides/recipes/examples/heroku-deployments.md#heroku-deployments